config/v1/types_infrastructure.go

```diff
@@ -620,9 +620,92 @@ const (
+
+	// IAMGCPServiceEndpointName is the name used for the GCP IAM Service endpoint.
+	IAMGCPServiceEndpointName GCPServiceEndpointName = "IAM"
+
+	// ServiceUsageGCPServiceEndpointName is the name used for the GCP Service Usage Service endpoint.
+	ServiceUsageGCPServiceEndpointName GCPServiceEndpointName = "ServiceUsage"
+
+	// StorageGCPServiceEndpointName is the name used for the GCP Storage Service endpoint.
+	StorageGCPServiceEndpointName GCPServiceEndpointName = "Storage"
+
+	// TagManagerGCPServiceEndpointName is the name used for the GCP Tag Manager Service endpoint.
+	TagManagerGCPServiceEndpointName GCPServiceEndpointName = "TagManager"
+)
+
+// GCPServiceEndpoint store the configuration of a custom url to
+// override existing defaults of GCP Services.
+type GCPServiceEndpoint struct {
+	// name is the name of the GCP service. This must be provided and cannot be empty.
+	// Allowed values are Compute, Container, CloudResourceManager, DNS, File, IAM, ServiceUsage,
+	// Storage, and TagManager.
+	// When set to Compute, the url field should be set to the private endpoint for the compute service. This
+	// overrides the default/public compute service endpoint.
+	// When set to Container, the url field should be set to the private endpoint for the container service. This
+	// overrides the default/public container service endpoint.
+	// When set to CloudResourceManager, the url field should be set to the private endpoint for the resource
+	// manager service. This overrides the default/public resource manager service endpoint.
+	// When set to DNS, the url field should be set to the private endpoint for the dns service. This
+	// overrides the default/public DNS service endpoint.
+	// When set to File, the url field should be set to the private endpoint for the file service. This
+	// overrides the default/public file service endpoint.
+	// When set to IAM, the url field should be set to the private endpoint for the iam service. This
+	// overrides the default/public iam service endpoint.
+	// When set to ServiceUsage, the url field should be set to the private endpoint for the service usage service. This
+	// overrides the default/public service usage service endpoint.
+	// When set to Storage, the url field should be set to the private endpoint for the storage service. This
+	// overrides the default/public storage service endpoint.
+	// When set to TagManager, the url field should be set to the private endpoint for the tag manager service. This
+	// overrides the default/public tag manager service endpoint.
+	//
+	// +required
+	Name GCPServiceEndpointName `json:"name"`
+
+	// url is fully qualified URI with scheme https, that overrides the default generated
+	// endpoint for a client.
+	// This must be provided and cannot be empty. The GCP maximum length is limited to 253 characters.
+	//
+	// +required
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:XValidation:rule="isURL(self)",message="must be a valid URL"
```

Comment:
It might make sense to include _what_ is considered a valid URL. In this case, it must be valid based on Go's net/url package interpretation of what a valid URL is https://pkg.go.dev/net/url#URL